fix(install): 清华 PyPI 镜像返回 HTTP 403 导致依赖安装失败

## 问题描述

执行 `./quickstart.sh --dev --yes` 时，pip 安装依赖包时从清华大学 PyPI 镜像（`https://pypi.tuna.tsinghua.edu.cn`）下载 `.whl` 文件时收到 **HTTP 403 Forbidden** 错误，导致多个依赖安装失败，最终整个安装流程以失败结束。

## 复现步骤

1. 克隆仓库并进入 SAGE 目录
2. 运行 `./quickstart.sh --dev --yes`
3. 安装过程中观察到多个依赖包报 403 错误

## 错误日志

```
Looking in indexes: https://pypi.tuna.tsinghua.edu.cn/simple/
Requirement already satisfied: setuptools>=68 in /opt/conda/lib/python3.12/site-packages (75.1.0)
Collecting setuptools>=68
  ERROR: HTTP error 403 while getting https://pypi.tuna.tsinghua.edu.cn/packages/.../setuptools-82.0.0-py3-none-any.whl
ERROR: Could not install requirement setuptools>=68 ...

ERROR: HTTP error 403 while getting https://pypi.tuna.tsinghua.edu.cn/packages/.../wheel-0.46.3-py3-none-any.whl
ERROR: Could not install requirement wheel>=0.42 ...

ERROR: HTTP error 403 while getting https://pypi.tuna.tsinghua.edu.cn/packages/.../wrapt-1.17.3-cp312-cp312-manylinux1_x86_64.whl
ERROR: Could not install requirement wrapt>=1.15.0,<2.0.0 ...

Collecting isage-common>=0.2.4.23
  ERROR: HTTP error 403 while getting https://pypi.tuna.tsinghua.edu.cn/packages/.../isage_common-0.2.4.24-py2.py3-none-any.whl
ERROR: Could not install requirement isage-common>=0.2.4.23 ...

[ERROR] 阶段失败: meta-package 依赖预安装 ✗
[ERROR] 阶段失败: 开发安装模式 ✗
```

## 问题分析

- 清华镜像索引页（`/simple/`）可以正常访问（返回 200），因此 pip 能找到包的元信息
- 但实际下载 `.whl` 文件时（`/packages/...`路径），服务器返回 **403 Forbidden**
- 这是清华镜像近期针对部分包的访问限制策略变更所致，并非代码本身的问题
- 受影响的包括：`setuptools`、`wheel`、`wrapt`、`isage-common` 等

## 影响

- `./quickstart.sh --dev --yes` 安装流程完全无法正常完成
- 在使用清华镜像的网络环境下（国内服务器/开发机）必现

## 建议修复方案

1. **多镜像自动回退**：`mirror_selector.sh` 在测速时仅测试 `/simple/` 端点可达性，应同时验证 `.whl` 文件下载能力（发起一次小文件 HEAD/GET 请求），将真正不可用的镜像排除
2. **依赖安装失败时自动切换镜像重试**：在 `core_installer.sh` 的安装逻辑中，当某个镜像返回 403 时，自动回退到下一个可用镜像重试，而非直接跳过
3. **短期 workaround**：在镜像列表中将清华镜像优先级降低，或在 CI/CD 及文档中说明遇到 403 时可手动指定其他镜像（如阿里云、华为云）

## 环境信息

- OS：Linux
- Python：3.12（/opt/conda）
- 受影响镜像：`https://pypi.tuna.tsinghua.edu.cn/simple`
- 时间：2026-03-02

--- a/tools/install/checks/mirror_selector.sh
+++ b/tools/install/checks/mirror_selector.sh
@@ -111,7 +111,8 @@ verify_mirror_download_capability() {
 
     local artifact_url
     artifact_url="$(extract_mirror_artifact_url "$mirror_url" "$test_package" 2>/dev/null || true)"
-    [ -n "$artifact_url" ] || return 0
+    # fail closed: 仅 /simple/ 可达不足以证明 wheel 下载可用
+    [ -n "$artifact_url" ] || return 1
 
     if ! command -v curl >/dev/null 2>&1; then
         return 1

--- a/tools/install/installers/environment_config.sh
+++ b/tools/install/installers/environment_config.sh
@@ -211,7 +211,8 @@ is_mirror_download_healthy() {
 
     local artifact_url
     artifact_url="$(extract_artifact_url_from_simple "$mirror_simple_url" "$test_package" 2>/dev/null || true)"
-    [ -n "$artifact_url" ] || return 0
+    # fail closed: /simple/ 可达但无法解析实际包链接时，不能判定该镜像可下载
+    [ -n "$artifact_url" ] || return 1
 
     local artifact_status
     artifact_status="$(probe_artifact_download_status "$artifact_url")"

--- a/tools/install/tests/test_e2e_integration.sh
+++ b/tools/install/tests/test_e2e_integration.sh
@@ -31,10 +31,10 @@ TEST_DIR="/tmp/sage_e2e_test_$$"
 # 导入颜色定义
 source "${SAGE_ROOT:-}/tools/install/ui/colors.sh"
 
-# 测试结果
+# 测试统计
 TOTAL_STEPS=0
-PASSED_STEPS=0
-FAILED_STEPS=0
+PASSED_CHECKS=0
+FAILED_CHECKS=0
 
 # 日志函数
 log_step() {
@@ -49,13 +49,13 @@ log_step() {
 log_success() {
     local message="$1"
     echo -e "${GREEN}✅ $message${NC}"
-    PASSED_STEPS=$((PASSED_STEPS + 1))
+    PASSED_CHECKS=$((PASSED_CHECKS + 1))
 }
 
 log_failure() {
     local message="$1"
     echo -e "${RED}❌ $message${NC}"
-    FAILED_STEPS=$((FAILED_STEPS + 1))
+    FAILED_CHECKS=$((FAILED_CHECKS + 1))
 }
 
 log_info() {
@@ -170,10 +170,14 @@ step4_test_venv_policy_reject() {
     source "$TEST_DIR/tools/install/ui/colors.sh"
     source "$TEST_DIR/tools/install/installers/environment_config.sh"
 
-    export VIRTUAL_ENV="$TEST_DIR/.venv"
-    if check_virtual_environment_isolation "pip" >/dev/null 2>&1; then
+    unset VIRTUAL_ENV
+    # check_virtual_environment_isolation 在检测到 Python venv 时会直接 exit 1；
+    # 这里放到子 shell 中验证，避免中断整个集成测试进程。
+    if (
+        export VIRTUAL_ENV="$TEST_DIR/.venv"
+        check_virtual_environment_isolation "pip" >/dev/null 2>&1
+    ); then
         log_failure "Python venv 未被拒绝"
-        unset VIRTUAL_ENV
         return 1
     fi
     unset VIRTUAL_ENV
@@ -264,18 +268,19 @@ print_test_summary() {
     echo -e "${BLUE}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━${NC}"
     echo -e "${BLUE}${BOLD}测试总结${NC}"
     echo -e "${BLUE}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━${NC}"
+    local total_checks=$((PASSED_CHECKS + FAILED_CHECKS))
     echo -e "总步骤数: $TOTAL_STEPS"
-    echo -e "${GREEN}通过: $PASSED_STEPS${NC}"
-    echo -e "${RED}失败: $FAILED_STEPS${NC}"
+    echo -e "${GREEN}通过检查项: $PASSED_CHECKS${NC}"
+    echo -e "${RED}失败检查项: $FAILED_CHECKS${NC}"
 
     local pass_rate=0
-    if [ $TOTAL_STEPS -gt 0 ]; then
-        pass_rate=$((PASSED_STEPS * 100 / TOTAL_STEPS))
+    if [ $total_checks -gt 0 ]; then
+        pass_rate=$((PASSED_CHECKS * 100 / total_checks))
     fi
-    echo -e "通过率: ${pass_rate}%"
+    echo -e "检查通过率: ${pass_rate}%"
     echo ""
 
-    if [ $FAILED_STEPS -eq 0 ]; then
+    if [ $FAILED_CHECKS -eq 0 ]; then
         echo -e "${GREEN}${BOLD}✅ 所有集成测试通过！${NC}"
         return 0
     else

--- a/tools/install/tests/test_environment_config.sh
+++ b/tools/install/tests/test_environment_config.sh
@@ -223,6 +223,51 @@ test_check_venv_rejected() {
 }
 
 # ============================================================================
+# 测试镜像下载探测
+# ============================================================================
+
+test_is_mirror_download_healthy_rejects_missing_artifact_url() {
+    echo ""
+    echo -e "${BLUE}测试组: is_mirror_download_healthy - 缺少 artifact URL 时应失败${NC}"
+
+    if (
+        curl() {
+            printf '200'
+        }
+        extract_artifact_url_from_simple() {
+            return 1
+        }
+
+        is_mirror_download_healthy "https://mirror.example/simple" "pip" >/dev/null 2>&1
+    ); then
+        assert_failure "缺少 artifact URL 时不应判定镜像健康"
+        false
+    else
+        assert_success "缺少 artifact URL 时正确拒绝镜像"
+    fi
+}
+
+test_verify_mirror_download_capability_rejects_missing_artifact_url() {
+    echo ""
+    echo -e "${BLUE}测试组: verify_mirror_download_capability - 缺少 artifact URL 时应失败${NC}"
+
+    if (
+        source "${SAGE_ROOT:-}/tools/install/checks/mirror_selector.sh"
+
+        extract_mirror_artifact_url() {
+            return 1
+        }
+
+        verify_mirror_download_capability "https://mirror.example/simple" "pip" >/dev/null 2>&1
+    ); then
+        assert_failure "mirror_selector 缺少 artifact URL 时不应判定可下载"
+        false
+    else
+        assert_success "mirror_selector 缺少 artifact URL 时正确拒绝镜像"
+    fi
+}
+
+# ============================================================================
 # 测试报告
 # ============================================================================
 
@@ -271,6 +316,8 @@ main() {
     test_check_venv_skip_in_ci
     test_check_venv_skip_for_conda_install
     test_check_venv_rejected
+    test_is_mirror_download_healthy_rejects_missing_artifact_url
+    test_verify_mirror_download_capability_rejects_missing_artifact_url
 
     # 打印总结
     print_test_summary

